gazenot/src/client.rs

```diff
@@ -0,0 +1,672 @@
+
+        Ok(Self(Arc::new(GazenotInner {
+            api_server: API_SERVER.to_owned(),
+            hosting_server: HOSTING_SERVER.to_owned(),
+            owner,
+            source_host,
+            auth_headers,
+            client,
+        })))
+    }
+
+    /// Ask The Abyss to create new ArtifactSets for the given packages
+    pub async fn create_artifact_sets(
+        &self,
+        packages: impl IntoIterator<Item = PackageName>,
+    ) -> Result<Vec<ArtifactSet>> {
+        // Spawn all the queries in parallel...
+        let mut queries = Vec::new();
+        for package in packages {
+            // Abyss is just an Arc wrapper around the real client, so Cloning is fine
+            let handle = self.clone();
+            let desc = format!(
+                "create hosting for {}/{}/{}",
+                self.source_host, self.owner, package
+            );
+            let url = self
+                .create_artifact_set_url(&package)
+                .map_err(|e| GazenotError::new(&desc, e))?;
+            queries.push((
+                desc,
+                url.clone(),
+                tokio::spawn(async move { handle.create_artifact_set(url, package).await }),
+            ));
+        }
+
+        // Then join on them all
+        join_all(queries).await
+    }
+
+    /// Ask The Abyss to create a new ArtifactSets for the given package
+    async fn create_artifact_set(
+        &self,
+        url: Url,
+        package: PackageName,
+    ) -> ResultInner<ArtifactSet> {
+        // No body
+        let response = self
+            .client
+            .post(url.clone())
+            .headers(self.auth_headers.clone())
+            .send()
+            .await?;
+
+        // Process the response
+        let ArtifactSetResponse {
+            public_id,
+            set_download_url,
+            upload_url,
+            release_url,
+            announce_url,
+        } = process_response(response).await?;
+
+        // Add extra context to make the response more useful in code
+        Ok(ArtifactSet {
+            package,
+            public_id,
+            set_download_url,
+            upload_url,
+            release_url,
+            announce_url,
+        })
+    }
+
+    /// Upload files to several ArtifactSets
+    ///
+    /// The input is a list of files to upload, but with each file parented
+    /// to the ArtifactSet it should be uploaded to.
+    ///
+    /// This is a bit of an awkward signature, but it lets us handle all the parallelism for you!
+    pub async fn upload_files(
+        &self,
+        files: impl IntoIterator<Item = (&ArtifactSet, Vec<Utf8PathBuf>)>,
+    ) -> Result<()> {
+        // Spawn all the queries in parallel...
+        let mut queries = vec![];
+        for (set, sub_files) in files {
+            for file in sub_files {
+                let handle = self.clone();
+                let filename = file.file_name().unwrap();
```

Comment:
Is this safe at this point?

Reply:
I should probably be a bit more paranoid about it for good errors, yeah